src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,124 @@
+
+using namespace chip;
+using namespace chip::System;
+using namespace chip::TLV;
+using namespace chip::Crypto;
+using namespace chip::SetupPayload;
+using namespace chip::Encoding::LittleEndian;
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateAdditionalDataPayload(uint16_t lifetimeCounter, char * serialNumberBuffer,
+                                                                         size_t serialNumberBufferSize,
+                                                                         PacketBufferHandle & bufferHandle,
+                                                                         bool enableRotatingDeviceId)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    TLVWriter innerWriter;
+    char rotatingDeviceIdBuffer[ROTATING_DEVICE_ID_MAX_LENGTH * 2 + 1];
+    size_t rotatingDeviceIdBufferSize = 0;
+
+    // Initialize TLVWriter
+    writer.Init(chip::System::PacketBufferHandle::New(chip::System::kMaxPacketBufferSize));
+
+    err = writer.OpenContainer(AnonymousTag, kTLVType_Structure, innerWriter);
+    SuccessOrExit(err);
+
+    if (enableRotatingDeviceId)
+    {
+        // Generating Device Rotating Id
+        err = generateRotatingDeviceId(lifetimeCounter, serialNumberBuffer, serialNumberBufferSize, rotatingDeviceIdBuffer,
+                                       ROTATING_DEVICE_ID_MAX_LENGTH * 2 + 1, rotatingDeviceIdBufferSize);
+        SuccessOrExit(err);
+
+        // Adding the rotating device id to the TLV data
+        err = innerWriter.PutString(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer);
+        SuccessOrExit(err);
+    }
+
+    err = writer.CloseContainer(innerWriter);
+    SuccessOrExit(err);
+
+    err = writer.Finalize(&bufferHandle);
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceId(uint16_t lifetimeCounter, char * serialNumberBuffer,
+                                                                    size_t serialNumberBufferSize, char rotatingDeviceIdBuffer[],
+                                                                    size_t rotatingDeviceIdBufferSize,
+                                                                    size_t & rotatingDeviceIdBufferOutputSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    Hash_SHA256_stream hash;
+    uint8_t outputBuffer[ROTATING_DEVICE_ID_MAX_LENGTH];
+    uint8_t hashOutputBuffer[kSHA256_Hash_Length];
+    BufferWriter outputBuferWriter(&outputBuffer[0], ArraySize(outputBuffer));
+    size_t rotatingDeviceIdBufferIndex = 0;
+
+    VerifyOrExit(rotatingDeviceIdBufferSize >= ROTATING_DEVICE_ID_MAX_LENGTH * 2 + 1, err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    // Computing the Rotating Device Id
+    // RDI = Lifetime_Counter + SuffixBytes(HMAC_SHA256(Serial_Number + Lifetime_Counter), 16)
+
+    err = hash.Begin();
+    SuccessOrExit(err);
+
+    err = hash.AddData((uint8_t *) serialNumberBuffer, serialNumberBufferSize);
```

Comment:
This is not what the spec says to do.  It says to do `HMAC-SHA256(key, data)`, not `SHA256(key || data)`.  Those are not the same thing.

Reply:
For my own knowledge: what is the key for the HMAC for the rotating id?

Reply:
Well, according to the latest security threat model meeting, we're just suggesting a hashing algorithm but it's up to the manufacturer to change that in their implementation, i update the spec to reflect this flexibility, and here we're going to use SHA256 instead

Reply:
Per spec right now, the key is defined as:

> The unique ID SHALL consist of a randomly generated 128-bit or longer number and programmed during factory provisioning. The unique ID MUST be protected against reading or writing, and stay fixed during the lifetime of the device.

Whether this serial number bit fits those criteria is not 100% clear to me.

Reply:
There is an issue created according to the threat model TT feedback, the suggested algorithm is not mandatory anymore, i will send a PR to update the spec.